RMT Onewire Peripheral

DaneSlattery opened the pull request:
https://github.com/esp-rs/esp-idf-hal/issues/445 related

Vollbrecht commented on bindings.h:
```diff
@@ -0,0 +1,2 @@
+#include "ds18b20.h"
+#include "onewire_bus.h"
```
Two points:
- if we introduce the usage of the onewire_bus.h in our api than we can directly include them in esp-idf-sys, and thous not needing the extra binding.h. That will than be conditionally included if the onewire component is included. I will update the changes for you in esp-idf-sys. 
- Lets focus in this PR on the onewire-bus itself first, so only parts of it are included and not subsequent other stuff.

DaneSlattery replied:
I will hide the onewire_bus behind an `#ifdef`. We can remove the ds18b20 for now.

I welcome you to update the `bindings.h` in esp-idf-sys, or I can make a related PR.

DaneSlattery replied:
I struggled to get this to work, with 
```
#ifdef ESP_IDF_COMP_ESPRESSIF__ONEWIRE_BUS_ENABLED
#include "onewire_bus.h"
#include "onewire_device.h"
#endif 
```

The `onewire` module in `bindings.rs` would be empty. Any ideas?

Vollbrecht replied:
I included it in esp-idf-sys for now, can you check if that is working for you. The module itself is optional. It did work for you without specifying the module before or did you always used the module?

DaneSlattery replied:
Done. It previously would only populate code inside the module because of the `Cargo.toml` `bindings_module` option. But inside `bindings.h`, the ESP_IDF_COMP_ESPRESSIF__ONEWIRE_BUS_ENABLED would always evaluate to 0 in my local setup. It works now with `bindings.h` in the `esp-idf-sys` crate, but now we're missing `onewire_device.h`.

Vollbrecht commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,42 @@
+{
```
at the end when the PR is finished remove this file.

DaneSlattery replied:
Will do. Might be good to add this path to gitignore

DaneSlattery replied:
Done

Vollbrecht commented on src/peripherals.rs:
```diff
@@ -10,6 +10,7 @@ use crate::mac;
 use crate::modem;
 #[cfg(any(esp32, esp32s2, esp32s3, esp32c6))]
 use crate::pcnt;
+#[cfg(feature = "legacy_rmt")]
```
if you are introducing this feature, you also need to first define it inside the `Cargo.toml` similar how it was done for the adc-oneshot-legacy.

DaneSlattery replied:
Done

DaneSlattery replied:
Done

Vollbrecht commented on src/onewire.rs:
```diff
@@ -0,0 +1,102 @@
+}
+
+/// wrapper around the device iterator to search for available devices on the bus
+pub struct DeviceSearch {
```
In your example you don't use the Device nor the DeviceSearch. If you create helper structs for you that you in the end only use internally yourself don't make them pub structs.

DaneSlattery replied:
I do use the public methods of these structs (eg `next_device()` and `address()`), so I think they should be exposed.

DaneSlattery replied:
I have annotated the types a bit clearer in the example to show where the public interface is

Vollbrecht commented on components_esp32.lock:
```diff
@@ -0,0 +1,21 @@
+dependencies:
```
we probably also want to add the component lock file to the .gitignore list

DaneSlattery replied:
Yeah that's a good idea

DaneSlattery replied:
Done

Vollbrecht reviewed (changes requested):
Thanks for working on this !

I left a couple notes, though overall i think first try to convince the CI, to let the project build for every target. If you succeeded at that point we can go into more details.

Vollbrecht commented:
You also get one additional cfg flag that you can use to gate the complete onewire bus implementation. When the remote component is included in the build system it will make the following available  #[cfg(ESP_IDF_COMP_ESPRESSIF__ESP32_CAMERA_ENABLED)] (in rust its all in lower case though ;D )

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,102 @@
+#[derive(Debug)]
+pub struct Device<'b> {
+    _device: onewire::onewire_device_t,
+    // _bus: &'b BusDriver<'b>, // not sure how I would hold a reference to this aside from the handle
```
@Vollbrecht
The internal esp-idf struct for `onewire_device_t` holds a handle to the onewire bus, and in rusty terms it means the `Device` lifetime is linked to the lifetime of the bus.

Does it make sense to include a reference to the bus wrapper so that the user can call `Device::bus(&self)` and get a `&BusDriver` return value? Or is it sufficient to just return the internal handle type?

DaneSlattery replied:
@ivmarkov any input?

ivmarkov replied:
Look at `SpiDeviceDriver`. Ideally, you need a _non_mutable_ borrow of the bus driver to be kept into `OWDeviceDriver`. For maximum flexibility (where the user either borrows a ref to the bus driver, or an `Arc` ref, or just moves an owned instance of the bus driver, use `T: Borrow<OWBusDriver>` - again - see `SpiDeviceDriver`.

ivmarkov replied:
But yes, you DO need to track the lifetime of the bus driver somehow (ideally- with a `Borrow` generic as per above). Else nothing stops the user from dropping the bus driver while the device driver is still alive, which is not ideal, to put it mildly.

DaneSlattery replied:
I attempted this, and it got pretty gnarly pretty quickly. https://gist.github.com/DaneSlattery/e0236dac98224bfe6f0137d75e209ca4

I don't know how to enforce that the device is dropped before the search is dropped, and my example does not compile anymore. I need to be able to drop a device without dropping the search.
```
error[E0597]: `search` does not live long enough
  --> examples/rmt_onewire.rs:68:23
   |
67 |     let mut search: DeviceSearch<OneWireBusDriver> = rmt_onewire.search()?;
   |         ---------- binding `search` declared here
68 |     let next_device = search.next_device();
   |                       ^^^^^^ borrowed value does not live long enough
...
80 | }
   | -
   | |
   | `search` dropped here while still borrowed
   | borrow might be used here, when `search` is dropped and runs the `Drop` code for type `DeviceSearch`
```

DaneSlattery replied:
This taught me a fair amount about lifetimes, thank you. I have implemented most of your suggestions. 

If I remove the phantomdata, I get this error about the lifetimes in `DeviceSearch` and `OWDevice`, even though the lifetime is captured in the constraint `T: Borrow<OWDriver<'b>>`

```
error[E0392]: lifetime parameter `'b` is never used
  --> src/onewire.rs:59:25
   |
59 | pub struct DeviceSearch<'b, T>
   |                         ^^ unused lifetime parameter
   |
   = help: consider removing `'b`, referring to it in a field, or using a marker such as `core::marker::PhantomData`

```

ivmarkov replied:
Given my comments below that `DeviceSearch` should actually NOT use a `T: Borrow<OWDevice<'b>>` (and by the way **sorry** if I wrongly seemed to hint you the other way in my comments above!!) - and given that `OWDevice` is likely no more, I'll resolve this.

As for your concrete problem at hand (if you face it in _other_ circumstances) - wouldn't the following work:
```rust
pub struct DeviceSearch<T> { ... }  // I.e. NO lifetime `'b` on the `DeviceSearch` struct itself

impl<'b, T> DeviceSearch<T>
where 
    T: Borrow<OWDevice<'b>>,
{
...
}
```
(Need to try that myself though, but don't see why it wouldn't work.)
?

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,102 @@
+    }
+}
+
+impl Drop for DeviceSearch {
```
I see this is done for other peripherals, do I need  to do it here as well?

DaneSlattery replied:
@ivmarkov

ivmarkov replied:
Not sure what is the question? Do you need to do _what_ w.r.t. `DeviceSearch`??

DaneSlattery replied:
Do I need to impl drop in this manner, as other peripherals have done?

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,102 @@
+        let pin = pin.into_ref().pin();
+        let bus_config = esp_idf_sys::onewire::onewire_bus_config_t { bus_gpio_num: pin };
+
+        let rmt_config = esp_idf_sys::onewire::onewire_bus_rmt_config_t { max_rx_bytes: 10 };
```
This could become configurable, but I want a working prototype first.

DaneSlattery commented on src/rmt.rs:
```diff
@@ -11,7 +11,7 @@
 //! Not supported:
 //! * Interrupts.
-//! * Receiving.
+//! * ~Receiving.~
```
This comment is outdated. I should also add something about how the legacy rmt is incompatible with the onewire rmt.

DaneSlattery commented on examples/rmt_onewire.rs:
```diff
@@ -0,0 +1,26 @@
+use esp_idf_hal::peripherals::Peripherals;
+use esp_idf_hal::rmt::{
+    FixedLengthSignal, PinState, Pulse, PulseTicks, Receive, RmtReceiveConfig, RmtTransmitConfig,
+    RxRmtDriver, TxRmtDriver,
```
unused

DaneSlattery reviewed (commented):
I've left some comments for myself and interested parties to weigh in.

DaneSlattery commented:
@Vollbrecht I think I need to also prevent the component from being included if the esp-idf version is 4, is there a flag for that?

DaneSlattery commented:
Struggling to get the v4 builds to go, I feel like it's an issue in the component inclusion that needs to be feature gated for versions <4

Vollbrecht commented:
> Struggling to get the v4 builds to go, I feel like it's an issue in the component inclusion that needs to be feature gated for versions <4, but I'm not 100% sure how to do that. I've gotten the v5 builds working across the board though, and I've tested the code is functioning locally on my `esp32`.

The core problem is that the external onewire_bus crate does not support esp-idf v4. And that is fine as we already deprecated the usage of such, so a new API doesn't need to also work there. It just shouldn't break the usage of the crate. The problem arises duo to you directly adding the onewire_bus inside the Cargo.toml of esp-idf-hal, and thous telling the buildsystem to include it, though it can not work with that crate version. 

The solution to the problem looks as follows. You are already using the cfg flag that is generated when `[[package.metadata.esp-idf-sys.extra_components]] ` is included. That is halve the problem solved. The rest is solved by not including the metadata section inside the default `Cargo.toml` of esp-idf-hal. By including it here it will always be pulled in, even if a user is not using the driver. The correct place for it to be included is at the binary level, e.g in the final binary a user is using.

With the inclusion of the headers in esp-idf-sys a user also will only need to add this:
```
[[package.metadata.esp-idf-sys.extra_components]]
remote_component = { name = "onewire_bus", version = "^1.0.2" }
```
and not specify any headers himself. 

For the example its a bit annoying. To run it you would have to provide  a modified Cargo.toml that includes it only for that example. Annoying but well i think without using something different than cargo itself unavoidable. You can add instructions in the doc header of the example how to run that example. It would look like `cargp run --example <example-name> --manifest-path <Cargo.toml that includes the metadata.esp-idf-sys section> `  In the future that potentially will be simpler for examples with the inclution of [cargo-script](https://rust-lang.github.io/rfcs/3424-cargo-script.html). 

TLDR: 
- put metdata.esp-idf-sys only in your final binary to signal that you want to use it
- for the example you need to pass a cargo.toml with the --manifest key path to a cargo.toml that includes it.

Vollbrecht commented on Cargo.toml:
```diff
@@ -65,3 +74,8 @@ esp-idf-sys = { version = "0.35", features = ["binstart"] }
+
+[[package.metadata.esp-idf-sys.extra_components]]
+remote_component = { name = "onewire_bus", version = "^1.0.2" }
+bindings_header = "bindings.h"
```
I merged the onewire_bus.h into esp-idf-sys. You can now remove this header stuff here and add a [patch.crates-io] section to poiint esp-idf-hal to espi.df-sys git master branch so you can test it out. E.g
```toml
[patch.crates-io]
esp-idf-sys = { git = "https://github.com/esp-rs/esp-idf-sys"}

```

DaneSlattery replied:
Done. Will remove the patch towards the end.

Vollbrecht commented on src/lib.rs:
```diff
@@ -46,7 +46,15 @@ pub mod peripheral;
 pub mod rmt;
+//
+#[cfg(all(
+    any(esp32, esp32c3, esp32c6, esp32h2, esp32p4, esp32s2, esp32s3),
```
since it works on any esp that has a rmt device, that maybe a better way to include it as to manually list all esp's. That can be done with the flag `esp_idf_soc_rmt_supported`.

DaneSlattery replied:
Curious where all these flags are listed/defined?

Vollbrecht replied:
`cargo rustc -- --print cfg` will show you what is currently defined for your combination of target and sdkconfig's

DaneSlattery commented:
@Vollbrecht unfortunately we need both includes. I've made a PR into esp-idf-sys. https://github.com/esp-rs/esp-idf-sys/pull/325

DaneSlattery commented:
I have an issue now in that the old RMT examples will never get compiled and may actually be broken. Perhaps we should add a CI stage to test with the `legacy_rmt` and `legacy_oneshot_adc` flags.

Vollbrecht commented:
> I have an issue now in that the old RMT examples will never get compiled and may actually be broken. Perhaps we should add a CI stage to test with the `legacy_rmt` and `legacy_oneshot_adc` flags.

its broken, because of the problem i stated in my last post. For idf v4 the old driver is always used, and don't need a flag because v4 only has the old driver. You can't include the onewire_bus inside v4, That's why it need to be moved from the esp-idf-hal `Cargo.toml` to the binary `Cargo.toml`

DaneSlattery commented:
> its broken, because of the problem i stated in my last post. For idf v4 the old driver is always used, and don't need a flag because v4 only has the old driver. You can't include the onewire_bus inside v4, That's why it need to be moved from the esp-idf-hal `Cargo.toml` to the binary `Cargo.toml`

Let me put it another way. Even for idf v5, the `rmt-legacy`  and `adc-oneshot-legacy` features are disabled in all of our tests because we don't do any tests with those features activated. We do a build for `no_std` and `alloc`, but not for these other features, which means we  aren't ensuring that these examples will still work.
![image](https://github.com/user-attachments/assets/a350d92e-dbc8-4962-8678-4748af79640f)

DaneSlattery commented:
🎉

Vollbrecht commented:
gj making the first milestone -> getting CI green ;D 

I am a bit busy right now, but will have a closer look at the weekend and hope to have some more feedback by then,

ivmarkov commented:
@Vollbrecht @DaneSlattery This had been going on for quite some time. I'm not sure in the meantime - is this ready for merge, and what (if any) are the outstanding issues?

DaneSlattery commented:
Hi @ivmarkov 

The build works and the code is tested. The outstanding items are this:

1. The rmt-legacy and adc-oneshot-legacy features are disabled in all of the CI/CD tests because we don't do any tests with those features activated. We do a build for no_std and alloc, but not for these other features, which means we aren't ensuring that these examples will still work.
2.  There are a few comments of mine, I will @ you on them

ivmarkov commented on src/onewire.rs:
```diff
@@ -0,0 +1,135 @@
+    _p: PhantomData<&'b ()>, // holds the lifetime since the device is linked to the lifetime of the bus
+}
+
+impl<'a> Device<'a> {
```
OK so how exactly do I send and receive data via the device driver?
I don't see a single method here for doing that?

DaneSlattery replied:
Essentially, a device is just an abstraction for address that we use, similar to the slave ID in I2C, except that it's static. Multiple devices exist on the OneWire bus, and each will require a further driver for their individual registers. I don't provide methods for reading/writing on raw devices because these `read` and `write`  methods are used by the underlying esp-idf C API for specific Devices. 

This does limit implementations to the C API only for now, but at least there is a way to expose them. They are 'extern` in `bindings.rs`. You may notice that none of these functions from the C API use the `onewire_device_t`, communication is still driven through the bus:

```
 pub fn onewire_bus_write_bytes(
        bus: onewire_bus_handle_t,
        tx_data: *const u8,
        tx_data_size: u8,
    ) -> esp_err_t;

 pub fn onewire_bus_read_bytes(
        bus: onewire_bus_handle_t,
        rx_buf: *mut u8,
        rx_buf_size: usize,
    ) -> esp_err_t;

 pub fn onewire_bus_write_bit(bus: onewire_bus_handle_t, tx_bit: u8) -> esp_err_t;

 pub fn onewire_bus_read_bit(bus: onewire_bus_handle_t, rx_bit: *mut u8) -> esp_err_t;

```


https://github.com/DaneSlattery/esp-idf-hal/blob/9443f23fd3bdac6b7023b1d4ae6ef6292ec19dc6/src/onewire.rs#L138  provides a starting point for the R/W. I will work on an example.

ivmarkov replied:
I do not understand why we should continue wrapping more and more C code - including the C code for the "specific" devices on the bus.

After all, all the additional C code does is to push and pull bits on the wire, likely using as you say read/write bit.

So why don't we just re-implement these additional drivers in Rust instead? There definitely won't be a performance slowdown because of that, because the most important methods - rerding and writing data (a single bit, really?) can be easily exposed and will become the onewire driver RW API.

What I'm trying to understand additionally, is whether there is a scheme which is device-independent as to how the master that speaks to a device "selects" the device to speak to. And how the master "broadcasts" to all devices. Can't imaging this to be device-dependent, because bits and pieces of this onewire thing, like the device "address" (the short and the long one) seem "standardized"?

ivmarkov replied:
It seems the IO interface of `OWBusDriver` should just be:
- `onewire_bus_write_bytes` - perhaps expose as `write(&mut self, data: &[u8]) -> Result<(), EspError>`?
- `onewire_bus_read_bytes` - perhaps expose as `read(&mut self, buf: &mut [u8]) -> Result<usize, EspError>` where the `usize` is how many bytes were read in the buffer?
- `onewire_bus_reset` - perhaps expose as `reset(&mut self) -> Result<(), EspError>`? (Not that I'm aware what a OW bus "reset" means.)

... and then I expect in the "bytes" above somehow to be an indication of for what device the data is written (for the "write" case) and what device sends us the data (for the "read" case).
I would have expected somehow the read/write methods to have a "device address" of sorts, but I don't see this in the C API. So maybe let me ask again: how do I _address_ a device via the onewire protocol? Can't believe this is device specific?

ivmarkov replied:
... as for exposing more C API - sure we can do that, but maybe not so ideal. Usually, none of the other protocols (SPI / I2C) work like that. You expose a generic way to read/write data over SPI and I2C and then you write another driver (usually in pure Rust) **on top** of the SPI / I2C protocol. This is e.g. how all the LED display drivers in Rust land work.

DaneSlattery replied:
> how do I address a device via the onewire protocol? Can't believe this is device specific?

From wikipedia:
1. The master starts transmission with a reset pulse, which pulls the wire low for a set time. This resets every slave device on the bus.
2. Every slave device shows that it exists by pulling the bus low for a set time.
3. The basic sequence of data is a reset pulse, followed by an 8-bit command, and then data is sent or received in groups of 8 bits.
4. Each device has a 64-bit serial number which can be used to address it. The LSB of the serial number is common amongst all devices from the same family.
5. There are several standard broadcast commands that all devices will implement. There are also commands to address a particular device. **The master can send a selection command, then the address of a particular device. The following command is then only executed by the addressed device**.

This is quite similar to the I2C protocol. Every device that is attached to the bus has a configured address, and will only respond when it is addressed.

ivmarkov replied:
Thanks for the pointer. The question is a bit to what extent to model the address of the device explicitly, with an extra parameter to - say - the write command d (not sure how reads work exactly). This is what the i2c driver does actually - https://github.com/esp-rs/esp-idf-hal/blob/master/src/i2c.rs#L276

DaneSlattery replied:
So it turns out that the middle 8 bytes of a message are used for the address when making a request. You can see this in the `rmt_onewire` example. I think transaction logic is a nice to have here, at least for this iteration?

ivmarkov replied:
I think the latest PR is OK in this regard, so I'll resolve this conversation.

ivmarkov commented on src/onewire.rs:
```diff
@@ -0,0 +1,135 @@
+    ///
+    /// The pin will be used as an open drain output.
+    pub fn new(
+        pin: impl Peripheral<P = impl crate::gpio::InputPin + crate::gpio::OutputPin> + 'd,
```
Since internally the ESP IDF onewire driver uses the new RMT driver (a single channel for TX/RX) (plus the pin thst you already track) you need to somehow track the usage of the RMT channel peripheral, as it is still a peripheral - just like the pin which you track already.

If you don't track it, nothing major would happen because the old and the new RMT driver cannot co-exist, yet would still be good to follow the pattern established with the other drivers.

Now, some of the "new" (ESP IDF 5+) C drivers no longer take a peripheral "id"/"number". For example, the old RMT driver used to take a channel number, but the new C RMT driver doesn't as it internally manages what channel to be used. If you try to instantiate more driver instances than hardware channels you have available, it would likely return a runtime `EspError`. So in a way - with the new RMT C driver you don't really know which concrete hardware channel peripheral will be used by that particular C driver instance.

With that said, _nothing stops us from continuing the old way and passing to the driver a RMT "peripheral"_ (note that you don't need "new" RMT channel peripherals for the new C RMT driver, because the peripherals are ghost structures which however model **actual** hardware on the chip and not higher-level C driver concepts).

So what I suggest is to just enhance the `new` constructor by passing an `impl ... Channel ...` peripheral that you can copy from the old RMT driver wrapper. While there is no warranty that the C driver will use _exactly_ the same channel that you pass to it - we don't care, as the channels are indistinguishable from each other. We just use - with the new C driver - the channels as a way of glorified "reference counting" of sorts.

DaneSlattery replied:
Yeah I had planned  to look into this. It becomes quite problematic as you mentioned below because I ifdef the whole rmt module.

ivmarkov replied:
Well I would not call it "quite problematic". #ifdef-ing only the driver inside `rmt` module is hopefully not a rocket science. Easiest way to do it is what I suggested for the examples:
* Define a `mod driver {}` internal module (not public)
* Move all code _except_ the `peripheral!` peripherals' definitions as well as the RMT peripheral traits (if any) _inside_ the driver module
* Put a `cfg` on top of the `driver `module
* Put a cfg-ed `pub use driver::*` in the `rmt` module

No big deal?

DaneSlattery replied:
No big deal. It's done.

ivmarkov commented on src/onewire.rs:
```diff
@@ -0,0 +1,135 @@
+        esp!(unsafe { onewire_bus_reset(self._bus) })
+    }
+
+    pub fn search(&mut self) -> Result<DeviceSearch, EspError> {
```
This is also problematic. In fact, I don't see how you can have an immutable borrow of the bus driver in the device driver, and still maintain the "iterator" approach for the `search` method. If you make this method `&self` rather than `&mut self`, it might work on the surface, _but_:
* I'm not sure the underlying C bus driver supports multiple concurrent device searches. Does it?
* You'll have a bit of a "device driver identity" problem here in that two subsequent searches will return two copies of the same device driver for the same device on the bus.

In fact, I'm not sure anymore that the device driver (what you currently call `Device`) is really a driver. I mean, it superficially resembles the SPI interface, in that in SPI you also have a bus + devices on it, however in the SPI case the devices are in no way dynamic - each device has a CS (chip select) pin, so it cannot just come and go. And in a way, with SPI - rather than "searching" for devices - you pre-create as many device drivers as you "know" up-front devices you have - each with its own CS pin.

Perhaps, whether the "device" is really a driver or just an "information" that a device might currently be available on the bus depends on how you send/receive data to the device.
Which goes up to my previous comment that I don't actually see in your code any methods to send/receive data...?

DaneSlattery replied:
I'm not married to the immutable borrow of the `bus` driver in the `device`. I am actually hoping to capture the relationship just using lifetimes.

I don't think concurrent device searches make sense, even if they are supported. Here is the doc-string
```
 Create an iterator to enumerate the 1-Wire devices on the bus\n\n
 @param[in] bus 1-Wire bus handle\n
 @param[out] ret_iter Returned created device iterator\n 
 @return\n      
- ESP_OK: Create device iterator successfully\n      
- ESP_ERR_INVALID_ARG: Invalid argument\n     
- ESP_ERR_NO_MEM: No memory to create device iterator\n     
- ESP_FAIL: Other errors"
```
, which doesn't mention anything about con-current searches. I think we should dis-allow them.

The "device" implementation is just information, you are correct. Every `OneWire` device has a unique identifier. Within that identifier is a "Family key" that allows the `OneWire` bus transmitter to identify the kind of device it is. When the C-API "finds" a device on the bus, it saves it's unique identifier into the `onewire_device_t` struct and returns it from the device search. 

That said, I don't think it is an issue for two subsequent searches to "find" the same device. While it's not typical, it is certainly possible for this to be hot-pluggable, and one might connect several devices to the bus, doing a search each time. Each of those search results is a valid device.

ivmarkov replied:
> I'm not married to the immutable borrow of the `bus` driver in the `device`. I am actually hoping to capture the relationship just using lifetimes.

I think some of the lifetime issues in your sample gist code come from this ^^^ statement. The purpose of the `Borrow` trait and the `T` generic is to avoid using direct `&` references to the bus driver in the first place. This way the user pays the price of dealing with the `T` generic when using the driver, but she is able to:
- Pass a regular `&bus_driver` reference at the place of `T`
- Pass a smart pointer reference (`Rc` or `Arc`) at the place of `T` (depending on whether the driver is `Sync` or not)
- Pass the whole driver by moving (i.e. without a reference)

This is what `Borrow` gives you. But then again, the whole point is moot, because we probably don't need `Device` to become a full-scale driver, but rather - just an informational struct without IO capabilities and `'static` interior.

> I don't think concurrent device searches make sense, even if they are supported. Here is the doc-string

Fortunately a moot point once `Device` becomes a simple data carrier object.

> The "device" implementation is just information, you are correct. Every `OneWire` device has a unique identifier. Within that identifier is a "Family key" that allows the `OneWire` bus transmitter to identify the kind of device it is. When the C-API "finds" a device on the bus, it saves it's unique identifier into the `onewire_device_t` struct and returns it from the device search.
> 
> That said, I don't think it is an issue for two subsequent searches to "find" the same device. While it's not typical, it is certainly possible for this to be hot-pluggable, and one might connect several devices to the bus, doing a search each time. Each of those search results is a valid device.

Yeah, we are going the "information" way it seems...

DaneSlattery replied:
> This is what Borrow gives you. But then again, the whole point is moot, because we probably don't need Device to become a full-scale driver, but rather - just an informational struct without IO capabilities and 'static interior.

If a device does not have IO capabilities, I imagine a "device specific driver" will need to wrap the bus without owning it, so that multiple devices can mutably use the bus. The device specific driver will also need to wrap the "simple data carrier `Device`" so that it knows which device it is addressing.

ivmarkov replied:
Why? Why can't the user just use the bus driver for sending / receiving data to a device? Why do we need a "device driver"?
- we do not have it in i2c which is the closest example
- we DO have it in spi, but there it is a necessity, as each device on the spi bus "owns" a cs pin (a peripheral)

DaneSlattery replied:
> Why can't the user just use the bus driver for sending / receiving data to a device?

They can, and it's completely fine for someone to do raw communication in this way. However, any device on the bus will have unique functions, error codes, registers and the like. The raw communication is not that useful on it's own, but this is more of a user concern than an `esp-idf-hal` concern.

ivmarkov replied:
How you'll model the "upper layer" drivers is completely up to you, and yes, you probably need at least a struct-per-device-type. But this _does not mean_ these would need each their own OW "device" driver. They would just use a shared bus driver.

ivmarkov replied:
@DaneSlattery I was re-reading the conversation and let me summarize, just to make sure we are on the same page:

> If a device does not have IO capabilities, I imagine a "device specific driver" will need to wrap the bus without owning it, so that multiple devices can mutably use the bus. 

Yes, this is a common problem. The `read`/`write`/`reset` methods on the bus driver will be `&mut` of course.
(
BTW, maybe we should just name it `OWDriver` as it is clear it would be a "bus" driver because there will be _no_ other OW driver in `esp-idf-hal` anyway. For I2C we did the same - `I2CDriver` is a "bus" driver, yet we just name it `I2CDriver`.
)

As for the common problem - we don't offer any native solution to this - it is up to the higher-level device-specific drivers to share the bus somehow - including when using the `I2C` driver. The easiest and natural way to do this of course is to put the `OWDriver` behind a `Mutex`, and e.g. the ["shared bus"](https://docs.rs/shared-bus/latest/shared_bus/) solution for I2C does exactly that.

We can implement something like `OWSharedBusDriver` (might even be a trait) that does it for the user, yet I feel we first need to nail down the main (non-shared) read/write/reset API.

> The device specific driver will also need to wrap the "simple data carrier Device" so that it knows which device it is addressing. I will work on an implementation of a device specific driver as an example.

This is where I'm a bit unclear what you mean. The device specific "driver" just needs some sort of **address**. And if I look at the API of the current `Device` structure you expose, it has just two items:
- A reference to the bus driver. This we absolutely don't need to be returned in the `Device` struct because it is _absolutely clear_ what bus we are talking about - this is the bus that returned the `Device` info after all, via its `search` method
- A `u64` address. This one we need, and this is maybe the only thing a device specific driver would need in addition to a (mutable reference to) the bus driver. So I would suggest:
  - A device specific "driver" takes on constructor (_usually_ - up to the device specific "Driver" implementor)
    - A `u64` that designates the device address
    - A mutable reference (or rather and best - a `T: BorrowMut<OWDriver<'a>>`) of the bus driver
  - A device specific "driver" does NOT therefore even need to know about the `Device` struct. It will be just an informational data carrier. See below.

Additionally:
- Let's turn the `Device` struct into a simple Rust newtype around `u64` (the address). Or even the `DeviceSearch` iterator could just return `u64`s - the addresses of the devices on the bus. 
- We don't really need the "bus driver" info which is currently in the `Device` struct. It is absolutely clear, on what bus this device is, simply because you call `search` on a **specific** bus driver

So something like:
```rust
impl<'a> OWDriver<'a> {
    pub fn new(...) {}

    pub fn search(&mut self) -> Result<impl Iterator<Item = u64>, EspError> {}

    pub fn reset(&mut self) -> Result<(), EspError> {}

    // Passing `None` for an address means broadcast the write to all devices connected to the bus
    pub fn write(&mut self, addr: Option<&u64>, data: &[u8]) -> Result<(), EspError> {}

    // Here I'm a bit unsure as I haven't checked in details how read actually works, and if this signature is good enough
    pub fn read(&mut self, addr: &u64, buf: &mut [u8]) -> Result<usize, EspError> {}
}
```

?

DaneSlattery replied:
I have just pushed an update implementing your changes. Let me know how you feel about the design I went with.

ivmarkov replied:
Yes - I think the design is good. Once you address the various nits (and remove `OWDevice` or describe why it is still necessary) I think we should be in a good shape for merging. Thanks a lot, and markind this as resolved.

ivmarkov commented:
> Hi @ivmarkov
> 
> The build works and the code is tested. The outstanding items are this:
> 
> 1. The rmt-legacy and adc-oneshot-legacy features are disabled in all of the CI/CD tests because we don't do any tests with those features activated. We do a build for no_std and alloc, but not for these other features, which means we aren't ensuring that these examples will still work.
> 2. There are a few comments of mine, I will @ you on them. I have not clicked "resolved" on any of @Vollbrecht comments, but I have done them.
> 3. I am not sure if this part needs to become a `peripheral``.
> 4. I need to create a follow-up PR using this OneWire interface for a real device.

I think at least a sketch of how the read/write API is needed - and I would say - as part of this PR rather than as a follow-up PR, because - depending how RW works (would you place the read/write methods on the "device" structure or on the "bus" structure?) this leads to one or another design.

The other big problem is how dynamic are the devices on the bus. Are they somehow static or can they come and go (say - by physically removing the device from the wire)? Even if they can come and go, how often does that happen?

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,135 @@
+        })
+    }
+    /// send reset pulse to the bus
+    pub fn reset_bus(&mut self) -> Result<(), EspError> {
```
I managed to do these without a mutable reference, so that the `OWDriver` can be borrowed into the `DeviceSearch` while still being resettable.

ivmarkov replied:
Yes, I agree that would work.

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+
+pub struct OneWireBusDriver<'d> {
```
I think `OWDriver` is fine

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,135 @@
+use crate::peripheral::Peripheral;
+
+#[derive(Debug)]
+pub struct Device<'b> {
```
This will no longer be a driver itself. However, it seems like a device can still implement this through borrowing the `OWDevice` reference to the bus:

```
pub struct DS18B20DeviceDriver<'a, T: Borrow<OWDriver<'a>>>(OWDevice<'a, T>);

impl<'a, T: Borrow<OWDriver<'a>>> DS18B20DeviceDriver<'a, T> {
    pub fn get_temp(&self) -> Result<f32, EspError> {
        self.0.bus().borrow().reset()?;

        self.ds18b20_send_command(Ds18b20Command::ConvertTemp as _);
        let mut buf = [0u8; 10];
        self.0.bus().borrow().read(&mut buf)?;
        println!("buffer = {buf:?}");
        let lsb = buf[0];
        let msb = buf[1];

        let temp_raw: u16 = (u16::from(msb) << 8) | u16::from(lsb);

        Ok(f32::from(temp_raw) / 16.0)
    }

    fn ds18b20_send_command(&self, cmd: u8) -> Result<(), EspError> {
        let mut buf = [0; 10];
        buf[0] = OWCommand::MatchRom as _;
        let addr = self.0.address().to_le_bytes();

        buf[1..9].copy_from_slice(&addr);
        buf[9] = cmd;
        println!(
            "seend command{cmd} to addr {} [{:?}], full_msg = {:?}",
            self.0.address(),
            addr,
            buf
        );
        let write_size = self.0.bus().borrow().write(&buf)?;
        // let write_size = bus.write(&buf)?;

        println!("Sent {write_size} bytes");

        Ok(())
    }
}
```

ivmarkov replied:
Or just 
```rust
pub struct DS18B20DeviceDriver<'a, T: Borrow<OWDriver<'a>>>(T);
```
?
I.e. no need for the `OWDevice` indirection, as it does not give us anything.

ivmarkov commented on src/lib.rs:
```diff
@@ -46,7 +46,21 @@ pub mod peripheral;
+#[cfg(all(feature = "rmt-legacy", esp_idf_comp_espressif__onewire_bus_enabled))]
+compile_error!("the onewire component cannot be used with the legacy rmt peripheral");
+
+#[cfg(any(feature = "rmt-legacy", esp_idf_version_major = "4"))]
```
This cannot be done so simply. As discussed during the previous iteration, you _do_ need the RMT peripherals which are defined by the `esp_idf_hal::rmt` module. And you need to pass a RMT channel to `OWDriver` in addition to the pin. So you can't just `#ifNdef` the `rmt` module when the `rmt-legacy` feature is not defined. What you need to do instead is to leave the `rmt` module here unconditionally, and then - _inside_ the `rmt` module - protect with an `#ifdef` all of the driver code, but leave the peripheral definitions always defined and valid - even when the legacy RMT driver is `#ifNdef`-ed.

DaneSlattery replied:
Done

ivmarkov commented on src/peripherals.rs:
```diff
@@ -10,6 +10,7 @@ use crate::mac;
 use crate::modem;
 #[cfg(any(esp32, esp32s2, esp32s3, esp32c6))]
 use crate::pcnt;
+#[cfg(any(feature = "rmt-legacy", esp_idf_version_major = "4"))]
```
As per above, remove the `cfg` line. The code in the `esp_idf_hal::rmt` module should be changed so that the RMT _peripherals_ are always defined. Even if the legacy RMT driver is `#ifNdef`-ed.

DaneSlattery replied:
Done

ivmarkov commented on src/peripherals.rs:
```diff
@@ -71,6 +72,7 @@ pub struct Peripherals {
     pub ledc: ledc::LEDC,
     #[cfg(esp32)]
     pub hledc: ledc::HLEDC,
+    #[cfg(any(feature = "rmt-legacy", esp_idf_version_major = "4"))]
```
Ditto.

DaneSlattery replied:
Done

ivmarkov commented on src/peripherals.rs:
```diff
@@ -168,6 +170,7 @@ impl Peripherals {
             ledc: ledc::LEDC::new(),
             #[cfg(esp32)]
             hledc: ledc::HLEDC::new(),
+            #[cfg(any(feature = "rmt-legacy", esp_idf_version_major = "4"))]
```
Ditto.

DaneSlattery replied:
Done

ivmarkov reviewed (changes requested):
@DaneSlattery I think we are getting there!
I've left a lot of comments, but these I hope would be quite easy to fix, as the latest approach seems solid.

My biggest request is to remove `OWDevice` completely, as I continue to fail to understand what purpose it serves? Please let me know if I'm missing something.

ivmarkov commented:
@DaneSlattery Unrelated to the implementation chosen in this PR, but people do seem to use UART successfully as well to implement onewire in hardware: https://developer.electricimp.com/resources/onewire

Interesting!

DaneSlattery commented on .gitignore:
```diff
@@ -6,3 +6,5 @@
 **/*.rs.bk
 ulp/ulp_start.o
 install-rust-toolchain.sh
+/.devcontainer
```
I think this change is good, prevents future accidental commits of the devcontainer

ivmarkov replied:
Fair enough.

DaneSlattery commented on .cargo/config.toml:
```diff
@@ -1,6 +1,6 @@
 [build]
-target = "riscv32imc-esp-espidf"
-#target = "xtensa-esp32-espidf"
+# target = "riscv32imc-esp-espidf"
```
Done

DaneSlattery commented on .cargo/config.toml:
```diff
@@ -1,6 +1,6 @@
-target = "riscv32imc-esp-espidf"
-#target = "xtensa-esp32-espidf"
+# target = "riscv32imc-esp-espidf"
+target = "xtensa-esp32-espidf"
```
Done

DaneSlattery commented on Cargo.toml:
```diff
@@ -24,12 +24,17 @@ std = ["alloc", "esp-idf-sys/std"]
 nightly = []
 wake-from-isr = [] # Only enable if you plan to use the `edge-executor` crate
-embassy-sync = [] # For now, the dependecy on the `embassy-sync` crate is non-optional, but this might change in future
+embassy-sync = [
```
Done

DaneSlattery commented on Cargo.toml:
```diff
@@ -42,13 +47,17 @@ libstart = ["esp-idf-sys/libstart"]
 embedded-can = "0.4.1"
 embedded-hal = "1"
-embedded-hal-0-2 = { package = "embedded-hal", version = "0.2.7", features = ["unproven"] }
+embedded-hal-0-2 = { package = "embedded-hal", version = "0.2.7", features = [
```
Done

DaneSlattery commented on Cargo.toml:
```diff
@@ -42,13 +47,17 @@ libstart = ["esp-idf-sys/libstart"]
 embedded-io-async = "0.6"
 esp-idf-sys = { version = "0.35", default-features = false }
-critical-section = { version = "1.1.1", optional = true, features = ["restore-state-none"] }
+critical-section = { version = "1.1.1", optional = true, features = [
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+use crate::peripheral::Peripheral;
+
+#[derive(Debug)]
+pub struct OWDevice<'b, T>
```
I like option 2. I also like implementing the device driver on `OWDevice` as I have shown in previous comments, but let's keep it simple.

DaneSlattery replied:
Done. I went with option 1 so I could also implement the command code method on the wrapper.

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+{
+    _search: onewire_device_iter_handle_t,
+    _bus: T,
+    _p: PhantomData<&'b ()>,
```
> Are you allowed to have - at any point in time - TWO onewire_device_iter_handle_t instances which can be iterated-over independently from each other - or is this not allowed?

I don't know, but I don't think we should allow it from an API perspective, it's not particularly useful

DaneSlattery replied:
Done

DaneSlattery replied:
I gave this a shot, see the example where I had to ensure the search was dropped before the bus can be borrowed immutably for communication

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+        Ok(buff.len())
+    }
+
+    pub fn write(&self, data: &[u8]) -> Result<usize, EspError> {
```
The underlying code writes all the bytes if the data is smaller than the configured buffer:
`        let rmt_config = esp_idf_sys::onewire_bus_rmt_config_t { max_rx_bytes: 10 };
`

DaneSlattery replied:
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+    Search = 0xf0,
+    MatchRom = 0x55,
+    SkipRom = 0xcc,
+    SearchAlarm = 0xec,
```
https://www.analog.com/en/resources/app-notes/1wire-search-algorithm.html . This code finds devices that are in an alarm state

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+    MatchRom = 0x55,
+    SkipRom = 0xcc,
+    SearchAlarm = 0xec,
+    ReadPowerSupply = 0xb4,
```
https://www.analog.com/en/resources/technical-articles/how-to-use-the-ds2482-isup2c-1wirereg-master.html

DaneSlattery commented on src/lib.rs:
```diff
@@ -46,7 +46,21 @@ pub mod peripheral;
+    not(esp_idf_version_major = "4"),
+    esp_idf_comp_espressif__onewire_bus_enabled,
+))]
+pub mod onewire;
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+where
+    T: Borrow<OWDriver<'b>>,
+{
+    _addr: u64,
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    /// Search for the next device on the bus and yield it.
+    fn next_device(&self) -> Result<u64, EspError> {
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+    _p: PhantomData<&'b ()>,
+}
+
+impl<'b, T> OWDevice<'b, T>
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+where
+    T: Borrow<OWDriver<'b>>,
+{
+    type Item = u64;
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Ok(dev) = self.next_device() {
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+        esp!(unsafe { onewire_del_device_iter(self._search) }).unwrap();
+    }
+}
+#[derive(Debug)]
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+#[derive(Debug)]
+
+pub struct OWDriver<'d> {
+    _bus: onewire_bus_handle_t,
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+
+pub struct OWDriver<'d> {
+    _bus: onewire_bus_handle_t,
+    _p: PhantomData<&'d ()>,
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+        esp!(unsafe { onewire_bus_reset(self._bus) })
+    }
+
+    pub fn search(&self) -> Result<DeviceSearch<impl Borrow<OWDriver<'_>>>, EspError> {
```
Done.

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+    // }
+}
+
+impl<'d> Drop for OWDriver<'d> {
```
Done

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+}
+
+// command codes
+#[repr(u8)]
```
Done

DaneSlattery commented on examples/rmt_onewire.rs:
```diff
@@ -0,0 +1,186 @@
+    esp_idf_comp_espressif__onewire_bus_enabled,
+))]
+fn ds18b20_send_command<'a>(
+    device: &OWDevice<'a, impl Borrow<OWDriver<'a>>>,
```
Done

DaneSlattery commented on examples/rmt_onewire.rs:
```diff
@@ -0,0 +1,186 @@
+    bus: &OWDriver,
+    cmd: u8,
+) -> Result<(), EspError> {
+    let mut buf = [0; 10];
```
Agreed. I added a todo!

DaneSlattery commented on src/onewire.rs:
```diff
@@ -0,0 +1,188 @@
+    SkipRom = 0xcc,
+    SearchAlarm = 0xec,
+    ReadPowerSupply = 0xb4,
+}
```
Done

ivmarkov commented:
@DaneSlattery No rush, but wondering if you still plan to fix the remaining issues?

ivmarkov commented:
Shall I again re-review? I see you have marked with "Done" most (if not all) of the outstanding issues?
Weird - I have NOT received an email notification for that... :)

ivmarkov commented:
Actually, no - there are still unaddressed outstanding issues...

DaneSlattery commented:
Hi Ivmarkov.
I have addressed most of the comments, but I am still planning to work on
1. reworking the affected examples so that there is only one ifdef
2. reworking the legacy RMT feature flag to allow for channel peripherals 

I am happy for you to mark as accepted the tasks that were `done` so far

ivmarkov commented:
> Hi Ivmarkov. I have addressed most of the comments, but I am still planning to work on
> 
> 1. reworking the affected examples so that there is only one ifdef
> 2. reworking the legacy RMT feature flag to allow for channel peripherals

Np - take your time. I just wasn't sure what is the status quo, and as mentioned I somehow did not receive notifications for the already addressed tasks.

> I am happy for you to mark as accepted the tasks that were `done` so far

I just did - thanks!

ivmarkov commented on src/onewire.rs:
```diff
@@ -0,0 +1,157 @@
+use crate::peripheral::Peripheral;
+
+/// Onewire Address type
+#[derive(Debug)]
```
Supernit:
- Also derive `Copy, Clone, Eq, PartialEq`, as this is just a Newtype over `u64` which is these things already

DaneSlattery replied:
Done

DaneSlattery commented on examples/rmt_morse_code.rs:
```diff
@@ -18,10 +18,13 @@ use esp_idf_hal::delay::Ets;
 use esp_idf_hal::gpio::*;
 use esp_idf_hal::peripheral::*;
 use esp_idf_hal::peripherals::Peripherals;
+#[cfg(any(feature = "rmt-legacy", esp_idf_version_major = "4"))]
```
done. Will do the other examples soon.

DaneSlattery commented on examples/rmt_onewire.rs:
```diff
@@ -0,0 +1,186 @@
+//! RMT Onewire Example
```
done

DaneSlattery commented on examples/rmt_neopixel.rs:
```diff
@@ -13,9 +13,12 @@ use anyhow::{bail, Result};
 use core::time::Duration;
 use esp_idf_hal::delay::FreeRtos;
 use esp_idf_hal::peripherals::Peripherals;
+#[cfg(any(feature = "rmt-legacy", esp_idf_version_major = "4"))]
```
done

DaneSlattery commented on examples/rmt_musical_buzzer.rs:
```diff
@@ -8,11 +8,13 @@ use core::time::Duration;
 
 use esp_idf_hal::delay::Ets;
 use esp_idf_hal::peripherals::Peripherals;
+#[cfg(any(feature = "rmt-legacy", esp_idf_version_major = "4"))]
```
done

DaneSlattery commented:
Done! I somehow don't think this CI failure is exactly my fault, not sure what's happening.

Vollbrecht commented:
> Done! I somehow don't think this CI failure is exactly my fault, not sure what's happening.

Thanks for continue working on this! Yeah currently they broke nightly for all users that are using the build-std feature though it seams a [patch](https://github.com/rust-lang/cargo/pull/14370) fixing it was already merged. So should hopefully working again tommorow :D

ivmarkov commented:
Yes let's wait until tmr when we would hopefully have a clean CI, which would ensure the latest changes in this PR do not introduce build errors we have to fix post-merge.

DaneSlattery commented:
Greeen!

ivmarkov commented:
Thank you for your persistence through this excercise. Merging.

Post merge, I might exchange the order of `channel` and `pin` in `OWDriver::new` because everything else has it with the peripheral first and pins following, but that's all there is to it that I see as a follow up. Great job!
